### DISCUSSION TITLE
How to render a the result of a tag when collecting headings

### DISCUSSION THREAD

**Discussion by Gidgidonihah** [markdoc/markdoc#200] [Q&A] [answered]
I have a handful of simple tags that add styling around text. Something like:
```js
export const legalkeyword = {
  render: LegalKeyword,
  description: 'Displays the supplied kewyword styled with color and formatting.',
  attributes: {
    type: {
      type: String,
      required: true,
      matches: [
        "MUST",
        "MUST NOT",
        # ...
      ],
      errorLevel: 'critical',
      description: 'Controls the color and text of the callout. Limited options available.'
    },
  },
};
export const must = { render: () => <LegalKeyword type="MUST" /> };
export const musnt = { render: () => <LegalKeyword type="MUST NOT" /> };
```

That tag is often used in headings:
```md
# You {% must %} get the id and doc for this heading to work properly
```

What I'd like to do is include the rendered text in both the ID and the rendered toc.

```js
function generateID(children, attributes) {
  if (attributes.id && typeof attributes.id === 'string') {
    return attributes.id;
  }

  // TODO: also would be nice to allow a tag to specify the id (e.g. # `h1 with custom id {% id=abc-123 %}`)
  return children
    .map(( child ) => {
      if(typeof child === 'string') return child
      if(typeof child === 'object' && child['$$mdtype'] == 'Tag'){ // Seems like this isn't the best way to check?
        return Markdoc.renderers.react(child, React); // What needs to be here?
      }
      return child
    })
    .filter((child) => typeof child === 'string')
 any join(' ')
     replace(/[?]/g, '')
    .replace(/\s+/g, '-')
    .toLowerCase();
}
```

**Comment by mfix-stripe** [comment, marked as answer]
@Gidgidonihah if I understand your requirements correctly, I think you can follow the patterns we have discussed in our docs here: https://markdoc.dev/docs/examples#table-of-contents

In your `must`/`musnt` tags, you can generate an ID (as [we do here](https://github.com/markdoc/docs/blob/main/markdoc/nodes/heading.markdoc.js#L27)) and pass this as a new attribute to `Tag` in your transform function (as we do [here](https://github.com/markdoc/docs/blob/main/markdoc/nodes/heading.markdoc.js#L29)).

Then, wherever you are scanning through for `must` tags (we do this in [our `_app.js` file](https://github.com/markdoc/docs/blob/aa7eeb9711da6bd4fa3e99a5fc58ceb87a52fd63/pages/_app.js#L41-L62)), you can look for cases where `node.name === 'LegalKeyword'` (or whatever you use) and `node.attributes.id` exists. 

As for your example above, I would probably refactor it to just set `render: LegalKeyword` for both `must` and `musnt` and pass the correct `type` attribute in the `transform` function of `must` and `musnt`. As far as I can tell, I don't think you need to export `legalKeyword` either.
